src/algebra/squarefree.lean

```diff
@@ -295,3 +420,184 @@ squarefree_iff_irreducible_sq_not_dvd_of_exists_irreducible
+  { rw [bit0_eq_two_mul, nat.mul_div_right _ (dec_trivial:0<2)],
+    refine h dec_trivial (λ p pp dp, nat.succ_le_of_lt (lt_of_le_of_ne pp.two_le _)),
+    rintro rfl, exact nat.not_two_dvd_bit1 _ dp }
+end
+
+lemma squarefree_bit1 (n : ℕ) (h : squarefree_helper n 1) :
+  squarefree (bit1 n) :=
+begin
+  refine h dec_trivial (λ p pp dp, nat.succ_le_of_lt (lt_of_le_of_ne pp.two_le _)),
+  rintro rfl, exact nat.not_two_dvd_bit1 _ dp
+end
+
+lemma squarefree_helper_0 {k} (k0 : 0 < k)
+  {p : ℕ} (pp : nat.prime p) (h : bit1 k ≤ p) : bit1 (k + 1) ≤ p ∨ bit1 k = p :=
+begin
+  rcases lt_or_eq_of_le h with (hp:_+1≤_) | hp,
+  { rw [bit1, bit0_eq_two_mul] at hp, change 2*(_+1) ≤ _ at hp,
+    rw [bit1, bit0_eq_two_mul],
+    refine or.inl (lt_of_le_of_ne hp _), unfreezingI { rintro rfl },
+    exact nat.not_prime_mul dec_trivial (lt_add_of_pos_left _ k0) pp },
+  { exact or.inr hp }
+end
+
+lemma squarefree_helper_1 (n k k' : ℕ) (e : k + 1 = k')
+  (hk : nat.prime (bit1 k) → ¬ bit1 k ∣ bit1 n)
+  (H : squarefree_helper n k') : squarefree_helper n k :=
+λ k0 ih, begin
+  subst e,
+  refine H (nat.succ_pos _) (λ p pp dp, _),
+  refine (squarefree_helper_0 k0 pp (ih p pp dp)).resolve_right (λ hp, _),
+  subst hp, cases hk pp dp
+end
+
+lemma squarefree_helper_2 (n k k' c : ℕ) (e : k + 1 = k')
+  (hc : bit1 n % bit1 k = c) (c0 : 0 < c)
+  (h : squarefree_helper n k') : squarefree_helper n k :=
+begin
+  refine squarefree_helper_1 _ _ _ e (λ _, _) h,
+  refine mt _ (ne_of_gt c0), intro e₁,
+  rwa [← hc, ← nat.dvd_iff_mod_eq_zero],
+end
+
+lemma squarefree_helper_3 (n n' k k' c : ℕ) (e : k + 1 = k')
+  (hn' : bit1 n' * bit1 k = bit1 n)
+  (hc : bit1 n' % bit1 k = c) (c0 : 0 < c)
+  (H : squarefree_helper n' k') : squarefree_helper n k :=
+λ k0 ih, begin
+  subst e,
+  have k0' : 0 < bit1 k := bit1_pos (nat.zero_le _),
+  have dn' : bit1 n' ∣ bit1 n := ⟨_, hn'.symm⟩,
+  have dk : bit1 k ∣ bit1 n := ⟨_, ((mul_comm _ _).trans hn').symm⟩,
+  have : bit1 n / bit1 k = bit1 n',
+  { rw [← hn', nat.mul_div_cancel _ k0'] },
+  have k2 : 2 ≤ bit1 k := nat.succ_le_succ (bit0_pos k0),
+  have pk : (bit1 k).prime,
+  { refine nat.prime_def_min_fac.2 ⟨k2, le_antisymm (nat.min_fac_le k0') _⟩,
+    exact ih _ (nat.min_fac_prime (ne_of_gt k2)) (dvd_trans (nat.min_fac_dvd _) dk) },
+  have dkk' : ¬ bit1 k ∣ bit1 n', {rw [nat.dvd_iff_mod_eq_zero, hc], exact ne_of_gt c0},
+  have dkk : ¬ bit1 k * bit1 k ∣ bit1 n, {rwa [← nat.dvd_div_iff dk, this]},
+  refine @nat.min_sq_fac_prop_div _ _ pk dk dkk none _,
+  rw this, refine H (nat.succ_pos _) (λ p pp dp, _),
+  refine (squarefree_helper_0 k0 pp (ih p pp $ dvd_trans dp dn')).resolve_right (λ e, _),
+  subst e, contradiction
+end
+
+lemma squarefree_helper_4 (n k k' : ℕ) (e : bit1 k * bit1 k = k')
+  (hd : bit1 n < k') : squarefree_helper n k :=
+begin
+  cases nat.eq_zero_or_pos n,
+  { subst n, exact λ _ _, squarefree_one },
+  subst e,
+  refine λ k0 ih, irreducible.squarefree (nat.prime_def_le_sqrt.2 ⟨bit1_lt_bit1.2 h, _⟩),
```

Comment:
Don't have this open in an editor, but is `h` the result from `cases` above? Name it explicitly?